### ISSUE TITLE
Discussion: Sequence abort convention

### ISSUE THREAD

**Issue by ioana-blue** [apache/openwhisk#995]
@perryibm @rabbah and I discussed sequences. I'm summarizing the conclusions of the discussion.

User-visible aspects:
- A sequence will have an activation id
- The logs of a sequence are constructed by the OW runtime as follows: each line contains an activation id and the result (success, failure) for the activation id
- Entitlement check is performed at the moment of invocation for all actions (i.e., to invoke a sequence, the user has to have the rights to invoke each action in the sequence). There won't be any partial execution of sequences (e.g., first 2 actions in the sequence are executed as the user has rights, while the remaining ones are not if the user doesn't have the right to invoke the third action in the sequence).

Implementation internals:
- The current implementation with `pipe.js` in the invoker will stay in place but guarded by a flag for experimental/comparison reasons
- The logic of executing a sequence are moved to the controller. The controller starts an actor each time it needs to execute a sequence. The actor executes in blocking fashion each action by posting directly to Kafka.
- The logs for the sequence are constructed as explained above and stored in the activation record for the sequence.
- The annotation for the activation will contain a field kind that indicates this activation was a sequence

More comments from @rabbah 
- entitlement check for the sequence component should be done both at create/update and invoke time.
- note: the actor doesn't perform an entitlement check before invoking - it's the route handler that does; the actor simply executes the invokes, waits for responses, and then writes the activation record.
- the result of the sequence should be the result of the last action in the sequence.

**Comment by ioana-blue** [comment]
Note: add test with a sequence that refers to a sequence, e.g., `seq1 = A, B; seq2 = seq1, C`

**Comment by ioana-blue** [comment]
More notes from Slack conversations:

* For throttling purposes, an action that is executing as a result of a sequence invocation should not count as an additional action for the user. In other words, for a sequence with N actions,  only one should be taken into account for throttling purposes. Not 100% sure but this may happen automatically since throttling happens before getting to the part that executes an action as part of a sequence. Need to double check that.
* We need to limit the number of actions within a sequence. Since sequence can refer to sequences, some recursive count is needed.

**Comment by ioana-blue** [comment]
@perryibm @rabbah @sjfink 

What's our convention for an action within a sequence that returns some json that contains an `error` field? Does that stop the execution of the sequence? That's what I remember, but I wanted to double-check.

**Comment by rabbah** [comment]
An initial implementation was done in #1083. The plan to get the functionality merged into the controller will be structured in the following phases (hence this epic):

1. resolve all sequence components into fully qualified names on action create/update
1. impost a limit on sequence length (and disallow recursion)
1. add entitlement checks to verify access at create/update time to referenced actions
1. implement continuations in controller to process the action sequence

Additional considerations:
1. action timeout for sequence needs to account for remaining slack
1. sequence activation id must be referenced by component activation ids via `caused` property
1. agree on `start` and `end` semantics for the sequence activation (do they exist? or just a duration)

@ioana-blue others to add?

**Comment by perryibm** [comment]
We decided that an error in the sequence would stop downstream ones from running.  The idea though is that getting the activation for this composite would include some message indicating abnormal termination for the seq as a whole.

**Comment by sjfink** [comment]
I'd like to support ECA (event condition action) cleanly if we can.   

i.e. if C then foo.

I'd like to implement this as a sequence C -> foo, where C tests a boolean condition and passes through its parameters if true, and somehow aborts if false.    However the "abort" in this case would not really be an error, but a normal "false" condition.

Proposal: if an action returns "null", it is considered "success", but terminates a sequence.

**Comment by perryibm** [comment]
Null being "{}"?

**Comment by rabbah** [comment]
It must be since actions must return an object.

**Comment by sjfink** [comment]
Yes,   "{}".    We can have "return null" be sugar for "return {}" maybe.

**Comment by ioana-blue** [comment]
@rabbah There is one additional item that is not directly related to sequences, but it is useful to have before implementing sequences:

0. Refactor code to include method for package resolution (i.e., fully resolved the actual package that a binding refers to) and method for generating all entities that need to be checked during entitlement check (this is also related to bindings and recursive bindings). 

A couple of other details that have been mentioned somewhere else:
1. The logs of the sequence activation record will contain the activation ids of the actions that ran. 
2. The annotations of the sequence activation record will contain a field "kind" with the value "sequence" for an easy way of figuring out which activations come from sequences. 

We also need to decide on how we count sequences toward DoS limits, e.g., do we count the sequence execution as +1 or +N where N is the number of actions that run as a result of invoking a sequence.

**Comment by sjfink** [comment]
Notes from discussion with @psuter today:

It might be going to far to define the empty object {} as an abort condition for sequences, since it's easy to imagine lots of actions that run correctly with empty inputs.   This is especially true when you consider that we can specify default values for parameters.

Philippe reminded me that we already have a convention that if the return object contains a field called "error", we have a convention that this indicates erroneous completion and terminates the sequence.

Perhaps it would be better to use this kind of convention.   For example, perhaps we could further adopt the convention that if the value of "error" is `null`, then it aborts the sequence, but for the purposes of logging, this is not considered a real error.   Alternatively we could reserve another keyword (abort?  exit? stop?) ... however maybe we should avoid reserving more keywords since each such keyword increases the chances that someone uses the keyword by accident.

**Comment by ioana-blue** [comment]
+1 for piggy-backing on "error" `null`

**Comment by ioana-blue** [comment]
In discussion with @rabbah we realized that we need to be able to deal with the legacy code which stores the sequence components as strings in the database. I'll add migration tests for this case to make sure we don't have any breaking changes.

**Comment by rabbah** [comment]
In one of the earlier comments, @sjfink suggesting returning "null" to indicate action success but abort a sequence. This has the advantage that we don't have to use an "error" convention but violates the json object in/out abstraction.

Here's a proposal to admit "null" without changing the action's `Dictionary -> Dictionary` signature.

Suppose the signature of the action was `Option[Dictionary] => Option[Dictionary]`. Then one has to check the input to determine if it's defined or empty (that is, check if there is a value provided or none).

We can reconcile this with our current action signature by allowing actions to return a `Dictionary` or `null`, where the former is conceptually boxed to `Some[Dictionary]` and the latter is boxed to `None`. Then, we guarantee that when you invoke an action it's always `Some[Dictionary]` and never `None`. That is, if the argument is `None` your action is not invoked. In this way, the action is only allowed to map over the defined value and not the empty value. Another way to understand this is that an action receiving `Null` returns `Null` by definition.

From a composition point of view it's sound as in one can compose g(f(option)) as long as they only map on the option (in the Scala sense for example). From the sequence implementation point of view, the action returning `Null` aborts the rest of the sequence but the activation is recorded as having a successful status (and _not_ an error).

So the proposal (elaborating Steve's above) which is superior then using a convention on a reserved field:
1. allow `null` as a valid return from actions (note that `null` is a valid JSON value, not a JSON object)
2. modify the sequence scheduler so that it aborts execution (with status recorded as `success`) at the first action returning `null`
3. record the result as null in the activation record

In other words, I move to accept @sjfink's "Proposal: if an action returns "null", it is considered "success", but terminates a sequence."

**Comment by rabbah** [comment]
Close @ioana-blue?

**Comment by sjfink** [comment]
reopening to revive discussion on sequence abort semantics

**Comment by rabbah** [comment]
Use openwhisk composer.
